### DISCUSSION TITLE
Engrammer - Engram layout for programmers

### DISCUSSION THREAD

**Discussion by sunaku** [binarybottle/engram-layout#45] [Show and tell]
[Engrammer]: https://github.com/sunaku/engrammer

Hi Arno and Engram community,

After 1.25 years of using Engram exclusively, I recently needed to type on a standard laptop keyboard featuring the QWERTY layout.  I found myself struggling to locate symbols/punctuation marks (especially those on shifted keys) as I reluctantly resorted to hunt-and-peck.  However, I was pleasantly surprised to find that it was significantly easier to type certain punctuation marks used in programming (such as backtick, backslash, semicolon, and equals) because they were **_unshifted:_** directly available, without needing to press Shift! :open_mouth:

Having tasted the renewed ease of access to the aforementioned punctuation marks (used heavily in programming editors, such as Vim, and also more generally in application shortcuts such as Control-Equals for the "zoom in" operation), I could no longer return to Engram verbatim. :star_struck:  Instead, I was motivated to rectify the situation by harnessing the best of both worlds. Behold, [the *Engrammer* layout][Engrammer]! :nerd_face: 

Cheers.

# [Engrammer] - Arno's Engram layout for programmers

     `~ 1! 2@ 3# 4$  5%  6^  7& 8* 9( 0) [{ ]}
        bB yY oO uU  '"  ;:  lL dD wW vV zZ =+ \|
        cC iI eE aA  ,<  .>  hH tT sS nN qQ
        gG xX jJ kK  -_  /?  rR mM fF pP

This is a programmer-friendly variant of Arno's Engram 2.0 keyboard
layout that helps maintain cross-proficiency with standard keyboards:

* Shifted pairs are standard (e.g. quotes don't shift to parentheses).
* Semicolon is placed relative to comma and period just like standard.
* Equals and square brackets are placed near their standard locations.

In addition, application shortcuts that assume standard shifted pairs,
such as Control-Equals for the "zoom in" operation, now work properly.

To illustrate the differences between this layout, Engram, and QWERTY:

>*Legend:* Yellow is Engram; Blue is Engram-like; Pink is new.
>![Rendering of this layout on a standard 60% keyboard.](https://github.com/sunaku/engrammer/raw/main/layout.png)
>![Rendering of this layout on an ortholinear keyboard.](https://github.com/sunaku/engrammer/raw/main/ortho.png)

For example, here are my split ortholinear keyboards using this layout:
>![Photo of MoErgo Glove80](https://sunaku.github.io/moergo-glove80-keyboard-photograph.jpg)
>![Layout of MoErgo Glove80](https://sunaku.github.io/moergo-glove80-keyboard-base-layer.png)
>![Photo of Ergohaven Remnant](https://sunaku.github.io/ergohaven-remnant-keyboard-photograph.jpg)
>![Layout of Ergohaven Remnant](https://sunaku.github.io/ergohaven-remnant-keyboard-base-layer.png)

**Comment by binarybottle** [comment]
Thank you, Suraj -- I'm so happy that the Engram layout is serving you well!  

As a programmer myself, I was torn between how much I wanted the layout to reflect the needs of someone typing English prose versus for programming. Since I program primarily in Python, I hardly ever make use of the semicolon, backslash, or back tick, so your changes favor particular programming languages, which is fine, but I think it should state that a bit more clearly. 

I would love to dive in and think with you about how to possibly optimize this layout further for programmers of different languages, but I'm in deep on a project right now, so will take a look back later on.

All the best!

**Reply by sunaku** [reply]
Hi Arno,

Sorry -- I didn't mean to imply that Engram wasn't good for programming.  On the contrary, I've used it exclusively for so long and I really like its logical placement of symbols. :nerd_face::sparkles: However, I think that maintaining cross-proficiency with standard shifting may prove useful when Engram typists must inevitably hunt-and-peck on standard keyboards (on other people's computers).  It may also lower the barrier to entry for prospective typists who are hesitant about non-standard shifting.

I've found it acceptable to downgrade from Engram's logical (but non-standard) shifted symbols to standard shifting because, in practice, I often reach for a dedicated symbols layer anyway (pictured below) on my QMK-programmable Dactyl Manuform keyboard when I need to type a symbol while programming. :seedling:  This layer continues to evolve over time as I seldom rearrange it to better suit the systems and programming languages that I use most often.  For instance, this is what it looks like today:

![Dedicated Symbols Layer](https://user-images.githubusercontent.com/9863/187973879-717addbe-ca0e-4f13-b8f6-5bba4c36e392.png)

* Backslash is on the thumb, so I can `\` escape all other symbols without repositioning my hand.
* I use snake_case, so underscore `_`  is at the same spot as the most frequent English letter `e`.
* Strings are frequent, so all varieties of quotation marks are centrally located in the home block.
* Adjacent bigrams:
  * `~/` for home directory paths in UNIX.
  * `/#` for URI fragment separators.
  * `#!` for shebang lines in UNIX scripts.
  * `</` for closing tag in HTML and XML.
  * `|>` for pipe operator in Elixir language.
  * `.*` for regular expressions.
  * `*.*` for filesystem globs.
* Inward rolling bigrams:
  * `->` for thin arrows in C, C++, and Elixir languages.
  * `=>` for fat arrows in Perl, Ruby, and Elixir languages.
  * `+=` for accumulation assignment in C and many languages.
  * `-=` for negative accumulation assignment; as the above.
  * `!(` for negating a grouping in Boolean expressions.
  * `!$` for last argument of previous command in Bourne shell.
  * `$?` for exit status of previous command in Bourne shell.
* Vim editor support:
  * `:` on lower thumb (more like the palm really) for command mode.
  * `^` and `$` on the home row, for jumping to beginning/end of line.
  * `<` and `>` on the home block, for decreasing/increasing indentation.
  * `{` and `}` on the home block, for jumping to previous/next paragraph.
  * `%` for jumping to the matching/enclosing delimiter at/around cursor.

And so on: this is the result of several hundreds of layout iterations over the last 8+ years.

I look forward to your future research on this topic.  Thanks again for your consideration. :bow:

Cheers.

**Comment by binarybottle** [comment]
Cool -- I am glad you continue to give attention to optimizing to your particular programming needs.  I think it would be good to adopt a naming system for what may become a family of Engram(mer) layouts focusing on particular people or programming languages.  Perhaps something like Engram(mer)-sunaku-v1, Engram(mer)-PRE (for Perl, Ruby, and Elixir), or Engram(mer)-CREP (for C/++, Ruby, Elixir, and Perl)?

**Reply by sunaku** [reply]
Hi Arno,

I'm honored by the suggestion but I think that, in practice, symbol layers are too idiosyncratic and dynamic to be granted official titles such as those. :trophy:  Especially with the rise of hardware programmable keyboards today, one can readily craft their own symbol layer and tweak it endlessly at whim.  Thus I can't imagine that anyone (including me!) would want to use my particular symbol layer as it stands today.  At best, it can serve to inspire fellow programmers as they design their own, following [design principles like this](https://getreuer.info/posts/keyboards/symbol-layer/index.html).

For instance, the symbols layer I presented last time has since evolved modestly to better accommodate my needs today:
![symbols_layer_2022-09-26](https://user-images.githubusercontent.com/9863/192386728-51328a86-8f2a-4be9-bada-c76d1c0420a3.png)
* Easier thin/fat arrows and self-modifying assignments on bottom.
* Single quote placement same as Engram for muscle memory reuse.
* Easier pipe bar for UNIX shell; logical pairing of Boolean operations.

Cheers.

**Reply by binarybottle** [reply]
Thank you again for sharing!  You are right about programmable keyboards making it easy to experiment.  Thanks also for the symbol layout design principles link -- I hadn't run across that one.

**Reply by sunaku** [reply]
Hi Arno, I turned these musings into [a proper article](https://sunaku.github.io/dactyl-manuform-5x6-keyboard.html) featuring the "final" [symbol layer layout](https://sunaku.github.io/dactyl-manuform-5x6-keyboard.html#symbol-layer) that I have settled on. Cheers.

![symbol layer](https://sunaku.github.io/dactyl-manuform-5x6-keyboard-symbol-layer.png)

**Reply by sunaku** [reply]
Hi Arno,

I recently [switched to a new keyboard](https://sunaku.github.io/ergohaven-remnant-keyboard.html) that has 4 less keys (since there are 2 less keys in each thumb cluster).  Naturally, this has required me to re-imagine my symbol layer and I've iterated its layout to [this stable state](https://sunaku.github.io/ergohaven-remnant-keyboard.html#symbol-layer):

![symbol layer](https://sunaku.github.io/ergohaven-remnant-keyboard-symbol-layer.png)

I especially find it a joy to type the inward-rolling `();` which sits just above the `you` sequence in Engram 2.0.

Cheers.

**Reply by sunaku** [reply]
Hi Arno,

I recently [switched to the *MoErgo Glove80* ergonomic keyboard](https://sunaku.github.io/moergo-glove80-keyboard.html) (my *endgame* choice for the foreseeable future!) which has a lot more keys than before. :nerd_face:  Naturally, this required me to re-imagine my symbol layer again, finally settling on [this stable state](https://sunaku.github.io/moergo-glove80-keyboard.html#symbol-layer):

![symbol layer](https://sunaku.github.io/moergo-glove80-keyboard-symbol-layer.png)

Coming from relatively smaller keyboards (with 58 and 62 keys total), I was suddenly faced with a lot of extra keys available on the Glove80 😅 so I've filled the void with special sequences like `.*` and `*.*` (which is a delightful inward-outward swiping motion with the thumb), as well as redundant symbols (such as comma on the top row for function calls and also on the bottom row for Vim f/t jump repetition).  These are more for fun & convenience than serious optimization, I would say. 🤓

Cheers.

**Comment by binarybottle** [comment]
Wonderful!  Thank you, @sunaku!  I look forward to taking out some time to give this a close read.

**Comment by IrrenWirr** [comment]
🙋 I also wanted to share my customization of Engram v2

(I relegated the closing of pairs to the AltGr modifier because my text editor already close them automatically, but I kept the closing parenthesis because a useless decoration always bring a smile 💕 (I also excluded `<` & `>` because they are not exclusively used as pairs (they can be used as arrows or as less/more symbol)))

```
(bottom-left:  No    modifier,   top-right:     AltGr+Shift modifier
 top-left:     Shift modifier,   bottom-right:  AltGr       modifier)
```
```
geometry: ISO

base: |
  ┌─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┬─────┲━━━━━━━━━━┓
  │ | ∇ │ 7 ≈ │ 5 ≠ │ 3 ] │ 1 » │ 9 ⊂ │ 8 ⊃ │ 0 ∂ │ 2 } │ 4 × │ 6 ± │ ^ ∧ │ \ ¨ ┃          ┃
  │ & ∇ │ ~ ≈ │ = ≠ │ [ ] │ « » │ < ⊂ │ > ⊃ │ ∞ ∂ │ { } │ * × │ + ± │ % ∧ │ / ¨ ┃ ⌫        ┃
  ┢━━━━━┷━━┱──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┺━━┳━━━━━━━┫
  ┃        ┃ B β │ Y ꝭ │ O ω │ U υ │ ' ’ │ " ” │ L λ │ D δ │ W Δ │ V ∀ │ Z ζ │ $ ⇒ ┃       ┃
  ┃ ↹      ┃   β │   ꝭ │   ω │   υ │ ( ‘ │ ) “ │   λ │   δ │   Δ │   ∀ │   ζ │ # ⇒ ┃       ┃
  ┣━━━━━━━━┻┱────┴┬────┴┬────┴┬────┴┬────┴┬────┴┬────┴┬────┴┬────┴┬────┴┬────┴┬────┺┓  ⏎   ┃
  ┃         ┃ C ψ │ I ι │ E ε │ A α │ ; ∅ │ : · │ H Σ │ T τ │ S σ │ N ν │ Q Ω │ ` ∈ ┃      ┃
  ┃ ⇬       ┃   ψ │   ι │   ε │   α │ , ∅ │ . · │   Σ │   τ │   σ │   ν │   Ω │ @ ∈ ┃      ┃
  ┣━━━━━━┳━━┹──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┬──┴──┲━━┷━━━━━┻━━━━━━┫
  ┃      ┃ ▀ █ │ G φ │ X ξ │ J ∃ │ K ∫ │ _ ¬ │ ! ‽ │ R Π │ M μ │ F Ξ │ P π ┃               ┃
  ┃ ⇧    ┃ ■ ▄ │   φ │   ξ │   ∃ │   ∫ │ - ¬ │ ? ‽ │   Π │   μ │   Ξ │   π ┃ ⇧             ┃
  ┣━━━━━━┻┳━━━━┷━━┳━━┷━━━━┱┴─────┴─────┴─────┴─────┴─────┴─┲━━━┷━━━┳━┷━━━━━╋━━━━━━━┳━━━━━━━┫
  ┃       ┃       ┃       ┃ ␣ ␣                            ┃       ┃       ┃       ┃       ┃
  ┃ Ctrl  ┃ super ┃ Alt   ┃ ␣ ␣                            ┃ AltGr ┃ super ┃ menu  ┃ Ctrl  ┃
  ┗━━━━━━━┻━━━━━━━┻━━━━━━━┹────────────────────────────────┺━━━━━━━┻━━━━━━━┻━━━━━━━┻━━━━━━━┛
```

**Reply by binarybottle** [reply]
Thank you for sharing, @IrrenWirr!  Does this include all Greek characters and symbols?

**Reply by IrrenWirr** [reply]
No, it only has some (only a few lower case Greek characters are missing but almost every upper case is missing, in addition to missing every accent/sign/symbol/special UTF-8 Greek characters), because I don't plan to write in Greek, I will only use them as nice symbols for programming and math.

**Comment by willpuckett** [comment]
I switched to Engrammer last week. It's been much more comfortable writing JavaScript: I use arrow functions quite a bit () => {} and having the parenthesis on the center column was not feeling good with home row mods in particular. Writing import statements (again, in JS) seems to feel more natural as well. I was also having a hard time reaching the angle brackets in the standard Engram layout, which may be more related to my hobbit fingers than a general difficulty, but JSX feels much better with Engrammer for me.

It's also been a nice side benefit that my keycaps are now correct on my MacBook.

**Reply by binarybottle** [reply]
Great to hear, @willpuckett !

**Reply by sunaku** [reply]
Cool, I'm glad to hear it's working better for you. :+1:  Since you're using home row mods, I'm assuming you have a programmable keyboard?  If that's the case, you might like to set up a symbol layer for programming: [here is mine](https://sunaku.github.io/moergo-glove80-keyboard.html#symbol-layer) for example, which places `();` right above the "you" in Engram (it's delightful to type as an inward roll :heart_eyes:) and also has inward rolling `=>` fat and `->` thin arrows.

**Reply by willpuckett** [reply]
I do use a revxlp at my desk, but I primarily type on my laptop keyboard (MacBook Air). The revxlp runs zmk and timerless mods, and I use Karabiner simlayers to have some hrm functionality on the MacBook (only 1 hrm can be applied at a time, although they can be combined with other mods [wip repo](https://github.com/willpuckett/karabiner-config-engram)). 

I'm still working on getting full hrm combos working in Karabiner... It's been challenging with the remapping to Engram and won't ever feel as natural as my mech, but it's nice to try and be as similar as possible I think so switching is more natural. I also filed a bug report with Apple last week to initiate a discussion about including hrms as an option for users in System Settings. I'm not sure if Karabiner will ever have low enough level access to make timerless mods possible, but it seems like users of any layout could benefit.

**Comment by FishyLeaf57** [comment]
Hello, 

I am sorry if this has been explained already or justified in the two web pages that explain how this layout came to be. 
I have read it over a few times and couldn't work it out.

What is the impact of switching s and n on the engram layout so that they are the same as Dvorak? - I have spent a few months learning Dvorak and am looking to switch to Engram... - to make it easier, I wanted to switch to s and n so that I had less to learn. And since N is more common than S, granted marginally, I thought it would be okay. However, I can't work out or rather have no idea how to work out what impact that has on any sfbs or any other bits that were taken into account for its current placement. 

Anyway, 
Many Thanks

**Reply by binarybottle** [reply]
Hello, @FishyLeaf57!  

You are of course more than welcome to reconfigure key layouts to meet your particular needs. I used to use Dvorak, then switched to Colemak before taking matters into my own hands!

The reason why the N and S are ordered as they are in the Engram layout is because of Guiding criterion #5:

**Guiding criteria**
1.  Assign letters to keys that don't require lateral finger movements.
2.  Promote alternating between hands over uncomfortable same-hand transitions.
3.  Assign the most common letters to the most comfortable keys.
4.  Arrange letters so that more frequent bigrams are easier to type.
**5.  Promote little-to-index-finger roll-ins over index-to-little-finger roll-outs.**
6.  Balance finger loads according to their relative strength.
7.  Avoid stretching shorter fingers up and longer fingers down.
8.  Avoid using the same finger.
9.  Avoid skipping over the home row.
10. Assign the most common punctuation to keys in the middle of the keyboard.
11. Assign easy-to-remember symbols to the Shift-number keys.

If you look at Step #2 on https://engram.dev/ you'll see that NS is far more common than SN:

 TH 100,272,945,963  3.56% 
 ND  38,129,777,631  1.35%
 ST  29,704,461,829  1.05%
 NT  29,359,771,944  1.04%
 CH  16,854,985,236  0.60%
 **NS  14,350,320,288**   
 CT  12,997,849,406
 TR  12,006,693,396       
 RS  11,180,732,354   
 NC  11,722,631,112
 RT  10,198,055,461

**Reply by FishyLeaf57** [reply]
Hello @binarybottle 

Thanks so much for that explanation. 
Since my original post, I have started to do a lot of research based on yours, and I came to the same conclusion in the end. 

I spent a lot of time looking at key swaps and how those affected finger usage and consecutive finger use. - It seemed as though the increase in the same finger use meant that it was not worth swapping S and N for me, and I also looked at swapping X and G as I wanted to reduce that load on my pinky, but again that didn't work out... 

What I don't like about Dvorak is the very heavy use of my right pinky on the L key. I didn't mind the crossing over of columns with my index finger so much, but I acknowledge that it puts a lot of strain and slows you down, etc. This was my main reason for looking for an alternative; I only wish I had done this research three months ago; that would have saved me trying to unlearn Dvorak in favour of your layout now. 

Anyway, thank you very much for all your work on this... - I hope I can pick it up quickly and run with it. My goal is 120 to 150 words per minute with about 98% accuracy. 

35+ words per minute with 95% accuracy should allow me to make the switch to work full-time. - I was nearly here with Dvorak in 6 weeks. Now I am back to 5 to 6 wpm on Engram... Anyway, I will get there... 

If you don't mind, what's your wpm on Engram? 

Many Thanks.

**Reply by binarybottle** [reply]
Indeed, one of the reasons I quickly moved away from Dvorak was that it had rather uniform distribution across the fingers, including the weaker fingers. I used a lot of emacs back then and my pinkies went haywire. 

Others have posted their experiences adopting the Engram layout and their current typing speed above -- please check those out.

I wish you all the best in adopting the Engram layout!

**Comment by binarybottle** [comment]
Indeed, one of the reasons I quickly moved away from Dvorak was that it had rather uniform distribution across the fingers, including the weaker fingers. I used a lot of emacs back then and my pinkies went haywire.

Others have posted their experiences adopting the Engram layout and their current typing speed -- please check those out.

I wish you all the best in adopting the Engram layout!


Cheers,

@rno

Arno Klein<https://binarybottle.com/>, PhD (he)
MATTER Lab<https://matter.childmind.org/> Director
Healey Scholar
Sr Research Scientist
Child Mind Institute<https://www.childmind.org/>
347-577-2091
________________________________

**Comment by binarybottle** [comment]
See https://github.com/binarybottle/engram-2021/discussions/86#discussion-10126586 for an announcement of the new Engram approach/layouts/repo!
